src/query/exist_query.rs

```diff
@@ -0,0 +1,304 @@
+        let schema = enable_scoring.schema();
+        let Some((field, _path)) = schema.find_field(&self.field_name) else {
+            return Err(TantivyError::FieldNotFound(self.field_name.clone()));
+        };
+        let field_type = schema.get_field_entry(field).field_type();
+        if !field_type.is_fast() {
+            return Err(TantivyError::SchemaError(format!(
+                "Field {} is not a fast field.",
+                self.field_name
+            )));
+        }
+        Ok(Box::new(ExistsWeight {
+            field_name: self.field_name.clone(),
+        }))
+    }
+}
+
+/// Weight associated with the `ExistsQuery` query.
+pub struct ExistsWeight {
+    field_name: String,
+}
+
+impl Weight for ExistsWeight {
+    fn scorer(&self, reader: &SegmentReader, boost: Score) -> crate::Result<Box<dyn Scorer>> {
+        let fast_field_reader = reader.fast_fields();
+        let dynamic_columns: crate::Result<Vec<DynamicColumn>> = fast_field_reader
+            .dynamic_column_handles(&self.field_name)?
+            .into_iter()
+            .map(|handle| handle.open().map_err(|io_error| io_error.into()))
+            .collect();
+        let mut non_empty_columns = Vec::new();
+        for column in dynamic_columns? {
+            if !matches!(column.column_index(), ColumnIndex::Empty { .. }) {
+                non_empty_columns.push(column)
+            }
+        }
+        // TODO: we can optimizer more here since in most cases we will have only one index
+        if !non_empty_columns.is_empty() {
+            let docset = ExistsDocSet::new(non_empty_columns, reader.max_doc());
+            Ok(Box::new(ConstScorer::new(docset, boost)))
+        } else {
+            Ok(Box::new(EmptyScorer))
+        }
+    }
+
+    fn explain(&self, reader: &SegmentReader, doc: DocId) -> crate::Result<Explanation> {
+        let mut scorer = self.scorer(reader, 1.0)?;
+        if scorer.seek(doc) != doc {
+            return Err(does_not_match(doc));
+        }
+        Ok(Explanation::new("ExistsQuery", 1.0))
+    }
+}
+
+pub(crate) struct ExistsDocSet {
+    columns: Vec<DynamicColumn>,
+    doc: DocId,
+    max_doc: DocId,
+}
+
+impl ExistsDocSet {
+    pub(crate) fn new(columns: Vec<DynamicColumn>, max_doc: DocId) -> Self {
+        let mut set = Self {
+            columns,
+            doc: 0u32,
+            max_doc,
+        };
+        set.find_next();
+        set
+    }
+
+    fn find_next(&mut self) -> DocId {
+        while self.doc < self.max_doc {
+            if self
+                .columns
+                .iter()
+                .any(|col| col.column_index().has_value(self.doc))
+            {
+                return self.doc;
+            }
+            self.doc += 1;
+        }
+        self.doc = TERMINATED;
+        TERMINATED
+    }
+}
+
+impl DocSet for ExistsDocSet {
+    fn advance(&mut self) -> DocId {
+        self.doc += 1;
+        self.find_next()
+    }
+
+    fn size_hint(&self) -> u32 {
```

Comment:
We need an implementation and a unit test for  `DocSet::seek`?

The implementation is rather trivial.

Reply:
I think at the moment it will be pretty much the same as default implementation. We could optimize it a tiny bit but I would rather do that in #2166 where we can add `pub fn doc_set_from_index(column_index: ColumnIndex) -> Box<dyn DocSet>;` and then implement something like AndDocSet, which can then perhaps have a more optimized advanced.

Reply:
not at all no.

You can do 


```rust
 fn advance(&mut self) -> DocId {
    self.seek(self.doc + 1)
 }

 #[inline(always)]
 fn seek(&mut self, target: DocId) -> DocId {
        self.doc = target;
        self.find_next()
 }
```

Reply:
Oh, I totally misinterpreted what default seek is actually doing. I see your point now.